[appletls]: Use SecIdentityCreate() to avoid using the Mac keychain.

baulig opened the pull request:
## Reading Certificates from the Mac Keychain

Reading the private key from the keychain is a new feature introduced with
AppleTls on XamMac and iOS.  On Desktop Mono, this new feature has several
known issues and it also did not received any testing yet.  We go back to the old
way of doing things, which is to explicitly provide an `X509Certificate2` with a
private key.

## Keychain Dialog Popups

When using Xamarin.Mac or Xamarin.iOS, we try to search the keychain
for the certificate and private key.

On Xamarin.iOS, this is easy because each app has its own keychain.

On Xamarin.Mac, the `.app` package needs to be trusted via code-sign
to get permission to access the user's keychain.  [FIXME: I still have to
research how to actually do that.]  Without this, you will get a popup
message each time, asking you whether you want to allow the app to access
the keychain, but you can make these go away by selecting "Trust always".

On Desktop Mono, this is problematic because selecting "Trust always"
give the 'mono' binary (and thus everything you'll ever run with Mono)
permission to retrieve the private key from the keychain.

This code would also trigger constant keychain popup messages,
which could only be suppressed by granting full trust.  It also makes it
impossible to run Mono in headless mode.

## SecIdentityCreate

To avoid these problems, we are currently using an undocumented API
called `SecIdentityRef()` to avoid using the Mac keychain whenever a
`X509Certificate2` with a private key is used.

On iOS and XamMac, you can still provide the `X509Certificate` without
a private key - in this case, a keychain search will be performed (and you
may get a popup message on XamMac).

chamons commented on mcs/class/System/Mono.AppleTls/AppleCertificateHelper.cs:
```diff
@@ -38,12 +38,39 @@ public static SecIdentity GetIdentity (X509Certificate certificate)
 				return SecKeyChain.FindIdentity (secCert, true);
 			}
+#else
+			return null;
```
Are internal users of this API going to check for null or could this cause NRE?

baulig replied:
Null is a valid return value - the caller is responsible for throwing an exception.

chamons commented on mcs/class/System/Mono.AppleTls/Certificate.cs:
```diff
@@ -309,6 +328,17 @@ public SecKey (IntPtr handle, bool owns = false)
 				CFObject.CFRetain (handle);
 		}
 
+		/*
```
Should ^ ctor set this.owner explicitly?

baulig replied:
Hmm, we could also make `owner` the object that's `CFRelease()`d and set it to `IntPtr.Zero` if we don't "own" it.

chamons commented on mcs/class/System/Mono.AppleTls/Certificate.cs:
```diff
@@ -331,7 +361,67 @@ public void Dispose ()
+			if (owner != IntPtr.Zero) {
+				CFObject.CFRelease (owner);
+				owner = handle = IntPtr.Zero;
+			} else if (handle != IntPtr.Zero) {
```
It seems a tad weird to have this else in the Dispose. Not necessary wrong, but not expected.

chamons commented on mcs/class/System/Mono.AppleTls/Certificate.cs:
```diff
@@ -331,7 +361,67 @@ public void Dispose ()
+		{
+			var descriptorHandle = CFString.Create (descriptor);
+			if (descriptorHandle == null)
+				throw new InvalidOperationException ();
```
Is this the right exception?

spouliot replied:
If that method is used (I did not see it used in the diff) then it seems this should be a null check on `descriptor` to throw an `ArgumentNullException`

chamons commented on mcs/class/System/Mono.AppleTls/Certificate.cs:
```diff
@@ -331,7 +361,67 @@ public void Dispose ()
+
+				return new SecAccess (accessRef, true);
+			} finally {
+				descriptorHandle.Dispose ();
```
Could we use a using block here instead?

baulig replied:
Bad idea, but we're incorrectly doing it in so many other places.  We'd have to make all those `CFString.Create()` etc. functions throw instead of possibly returning null.

marek-safar replied:
The idea is not to change `CFString.Create()` but keep the code as it but replace try/finally with using

chamons commented on mcs/class/System/Mono.AppleTls/ImportExport.cs:
```diff
@@ -54,9 +62,168 @@ static public SecStatusCode ImportPkcs12 (CFData data, CFDictionary options, out
+
+			IntPtr result = IntPtr.Zero;
+			SecStatusCode status = SecStatusCode.Success;
+			for (var test = 0; test < 13; test++) {
```
Why 13 here?

baulig replied:
Oops - that entire loop must be removed!

chamons commented on mcs/class/System/Mono.AppleTls/ImportExport.cs:
```diff
@@ -54,9 +62,168 @@ static public SecStatusCode ImportPkcs12 (CFData data, CFDictionary options, out
+			using (var cert = new SecCertificate (certificate)) {
+				var identity = SecIdentityCreate (IntPtr.Zero, cert.Handle, key.Handle);
+				if (CFType.GetTypeID (identity) != SecIdentity.GetTypeID ())
+					throw new InvalidOperationException ();
```
Right exception here?

baulig replied:
It should never happen - do we have some kind of "internal error" exception?

marek-safar commented on mcs/class/System/Mono.AppleTls/ImportExport.cs:
```diff
@@ -54,9 +62,163 @@ static public SecStatusCode ImportPkcs12 (CFData data, CFDictionary options, out
+
+			IntPtr result = IntPtr.Zero;
+			SecStatusCode status = SecStatusCode.Success;
+			keyParams = new SecItemImportExportKeyParameters ();
```
What is this line for?

marek-safar commented on mcs/class/System/Mono.AppleTls/ImportExport.cs:
```diff
@@ -54,9 +62,163 @@ static public SecStatusCode ImportPkcs12 (CFData data, CFDictionary options, out
+				Marshal.StructureToPtr (keyParams.Value, keyParamsPtr, false);
+			}
+
+			IntPtr result = IntPtr.Zero;
```
You don't need to explicitly set there when they are set by SecItemImport

spouliot commented on mcs/class/System/Mono.AppleTls/Certificate.cs:
```diff
@@ -331,7 +361,67 @@ public void Dispose ()
+		[DllImport (AppleTlsContext.SecurityLibrary)]
+		extern static /* OSStatus */ SecStatusCode SecAccessCreate (/* CFStringRef */ IntPtr descriptor,  /* CFArrayRef */ IntPtr trustedList, /* SecAccessRef _Nullable * */ out IntPtr accessRef);
+
+		public static SecAccess Create (string descriptor)
```
where is that method used ?

baulig replied:
It will be used later on - after we re-enabled the keychain on mobile.

marek-safar commented on mcs/class/System/Mono.AppleTls/ImportExport.cs:
```diff
@@ -54,9 +62,163 @@ static public SecStatusCode ImportPkcs12 (CFData data, CFDictionary options, out
+		// Native enum; don't change.
+		enum SecExternalFormat : int {
+			Unknown = 0,
+			OpenSSL,
```
If it's native enum, set OpenSSL value explicitly

marek-safar commented on mcs/class/System/Mono.AppleTls/ImportExport.cs:
```diff
@@ -54,9 +62,163 @@ static public SecStatusCode ImportPkcs12 (CFData data, CFDictionary options, out
+
+		// Native enum; don't change.
+		enum SecExternalItemType : int {
+			Unknown,
```
Explicitly set values

spouliot commented on mcs/class/System/Mono.AppleTls/Enums.cs:
```diff
@@ -9,8 +9,386 @@ namespace Mono.AppleTls {
 	// values are defined in Security.framework/Headers/SecBase.h 
 	enum SecStatusCode {
 		Success 							= 0,
-		DuplicateItem	 					= -25299,
```
It is also already public in Xamarin.*.dll so we'll end up shipping it twice (lots of string metadata that only grows over time).

I understand it's nice for debugging the internals but it's not exposed. Maybe it could live under a `#if XXX` ?

baulig replied:
We could probably remove some of the more obscure values.

marek-safar commented on mcs/class/System/Mono.AppleTls/Items.cs:
```diff
@@ -98,39 +111,48 @@ static SecIdentity FindIdentity (Predicate<SecCertificate> filter)
+
+			IntPtr ptr;
+			result = SecItem.SecItemCopyMatching (query.Handle, out ptr);
+			if ((result == SecStatusCode.Success) && (ptr != IntPtr.Zero)) {
```
Extra parens

marek-safar commented on mcs/class/System/Mono.AppleTls/Items.cs:
```diff
@@ -98,39 +111,48 @@ static SecIdentity FindIdentity (Predicate<SecCertificate> filter)
+			IntPtr ptr;
+			result = SecItem.SecItemCopyMatching (query.Handle, out ptr);
+			if ((result == SecStatusCode.Success) && (ptr != IntPtr.Zero)) {
+				var array = CFArray.ArrayFromHandle<INativeObject> (ptr, p => {
```
why not just `return CFArray.ArrayFromHandle...`

baulig replied:
Not my code, I only moved it around a little bit.  I just made this private as it's only supported as called from FindIdentity().

In some situations, the `result` array may contain elements of different types.

marek-safar commented on mcs/class/System/Mono.AppleTls/Items.cs:
```diff
@@ -98,39 +111,48 @@ static SecIdentity FindIdentity (Predicate<SecCertificate> filter)
+					IntPtr cfType = CFType.GetTypeID (p);
+					if (cfType == SecCertificate.GetTypeID ())
+						return new SecCertificate (p, true);
+					else if (cfType == SecKey.GetTypeID ())
```
Redundant else(s)

baulig commented:
I did some cleanups, fixed all the problems you guys addressed and added some documentation.

Could you please have another look at this to see whether this can now go in?  @marek-safar @chamons

marek-safar commented on mcs/class/System/Mono.AppleTls/Items.cs:
```diff
@@ -40,13 +40,26 @@
+		internal SecKeyChain (IntPtr handle, bool owns = false)
+		{
+			if (handle == IntPtr.Zero)
+				throw new Exception ("Invalid handle");
```
Should be ArgumentException
